Documentation of the routing algorithm (#2688)
CanalTP/navitia: documentation/rfc/doc-routing-algorithm.md

# Documentation of the routing algorithm

## The implemented algorithm

### Abstract

a quick view

### Example 1

![transit_map](transit_map.png)

Line 1 (every 20 minutes):

| A  | E  |
|----|----|
|7:50|8:50|
|8:10|9:10|
|8:30|9:30|

Line 2 (every 10 minutes):

| A  | B  | C  | D  |
|----|----|----|----|
|8:00|8:30|8:33|8:36|
|8:10|8:40|8:43|8:46|
|8:20|8:50|8:53|8:56|

Line 3 (every 5 minutes):

| D  | C  | B  | E  | F  |
|----|----|----|----|----|
|8:35|8:38|8:41|9:00|9:02|
|8:40|8:43|8:46|9:05|9:07|
|8:45|8:48|8:51|9:10|9:12|

Line 4 (every 5 minutes):

| B  | C  | F  | E  |
|----|----|----|----|
|8:35|8:40|9:00|9:02|
|8:40|8:45|9:05|9:07|
|8:45|8:50|9:10|9:12|

Line 5 (every 20 minutes):

| E  | F  |  G  |
|----|----|-----|
|9:05|9:07|10:00|
|9:25|9:27|10:20|
|9:45|9:47|10:40|

Connections: only intra stop point, 2 minutes except for F 3 minutes.

> ### RAPTOR and the specificities

Reviewer: It would be nice to have a small part on Pareto objectives and how they are themselves lexicographic objectives.
This would help understand the "multi-objective on arrival time" in second pass' description.

Author: That's not on raptor, that's on the global objectives. Our implementation of raptor only optimize arrival datetime and number of transfers

> The routing algorithm is build around RAPTOR, a routing algorithm developped at Microsoft and published in 2012. The paper can be found [here](https://www.microsoft.com/en-us/research/wp-content/uploads/2012/01/raptor_alenex.pdf).

Reviewer: :gb: "is buil`t` around"

> Basically, you give to RAPTOR the set of journey patterns it can use, the set of stop points it can use (for transfers), the datetime you can reach the stop points and the direction. It gives you, for each stop points and each number of transfers, the earliest arrival datetime.

Reviewer 2: ```suggestion
Basically, you give RAPTOR a set of journey patterns it can use, a set of stop points (for transfers) and date-times to reach the stop points. It gives you, for each stop points and each number of transfers, the earliest arrival date-time.
```

> Our implementation is very close to the original RAPTOR algorithm. There is 2 added functionnalities:

Reviewer 2: ```suggestion
Our implementation is very close to the original RAPTOR algorithm. There are 2 added functionalities:
```

>  - stay in: they are oportunistic, i.e. it is not guananteed to be optimally using the stay in. The journey patterns doesn't take the stay in into account.

Reviewer 2: ```suggestion
 - stay in: they are opportunistic, i.e. using the 'stay in' is not guaranteed to be optimal. Journey patterns don't take 'stay in' into account.
```

>  - ITL: as we have only one zone per stop time, the implementation is quite straightforward: on the first pickup, we set the current zone to the zone of the stop point. If the zone changes, no more drop off is forbidden because we can choose the good pick up afterwards.

Reviewer: "ITL" is :fr: , it should be translated, or at least precise it's short for french  "Interdiction de Transport Local".
This deserves a bit more of reference to what it does I think.

>  - ITL: as we are only one zone per stop time, the implementation is quite straightforward: at the furst pickup, we set the current zone to the zone of the stop point. If the zone change, no more drop off is forbidden because of the zone because we can choose afterward the good pick up.

Reviewer 2: ```suggestion
 - ITL: as we have only one zone per stop time, the implementation is quite straightforward: on the first pickup, we set the current zone to the zone of the stop point. If the zone changes, no more drop off is forbidden because we can choose the good pick up afterwards.
```

Reviewer 2: TBH, I'm not quite sure I understand this ITL thing :confused:

>  Journey patterns are generated automatically. They must have the same succession of stop times (except date-times) and must not overtake.

Reviewer: Can't you just say "the same succession of  `stop point` " then?

Author: No, there is also pick up, drop off and accessibilities.

> ### The first pass on example 1

Reviewer: You should precise that we start from A (and maybe go to G).
Also precise that there may be multiple start points more explicitly.

Author: The starting point and its departure datetime is written on each table

> ### The first pass on example 1

Reviewer: Maybe you can keep the same structure than on second pass:
## First pass
### First pass on example 1

>  The journey pattern are generated automatically. They must have the same succession of stop times (except datetimes) and must not overtake.

Reviewer 2: ```suggestion
 Journey patterns are generated automatically. They must have the same succession of stop times (except date-times) and must not overtake.
```

> In this section, you can find some examples of the result of raptor. Our implementation doesn't remember the vehicle journeys used, they are chosen by the raptor solution reader, explained later.

Reviewer 2: ```suggestion
In this section, you can find raptor results examples. Our implementation doesn't remember the vehicle journeys used, they are chosen by the raptor solution reader, explained later.
```

> The input of the algorithm is a set of journey patterns and stop points that can be used, and a starting datetime plus durations to access the reachable stop points. These 2 last inputs are used to compute the first TR0 line.

Reviewer 2: ```suggestion
The algorithm's input is a set of journey patterns and stop points, along with a starting date-time plus durations to access the reachable stop points. These last 2 inputs are used to compute the first TR0 line.
```

> The PTx lines correspond to the earliest arrival to a stop point using x vehicles (including stay in). The TRx corresponds to the earliest arrival to a stop point after x vehicles plus the connections.

Reviewer: `PT` and `TR` are short for what?

Author: Public transport and Transfers, matching approximatively the code.

> The PTx lines correspond to the earliest arrival to a stop point using x vehicles (including stay in). The TRx corresponds to the earliest arrival to a stop point after x vehicles plus the connections.

Reviewer: :gb: earliest arrival `at` a stop point

> The PTx lines correspond to the earliest arrival to a stop point using x vehicles (including stay in). The TRx corresponds to the earliest arrival to a stop point after x vehicles plus the connections.

Reviewer: this "including stay in" is ambiguous to me: stay in does increment 'x' ?

Author: no

> The PTx lines correspond to the earliest arrival to a stop point using x vehicles (including stay in). The TRx correspond to the earliest arrival to a stop point after x vehicles plus the connections.

Reviewer 2: ```suggestion
The PTx lines correspond to the earliest arrival to a stop point using x vehicles (including stay in). The TRx corresponds to the earliest arrival to a stop point after x vehicles plus the connections.
```

Starting from A at 7:45:

|level| A  | B  | C  | D  | E  | F  |  G      |
|-----|----|----|----|----|----|----|---------|
|TR0  |7:45|    |    |    |    |    |         |
|PT1  |    |8:30|8:33|8:36|8:50|    |         |
|TR1  |    |8:32|8:35|8:38|8:52|    |         |
|PT2  |    |    |    |    |    |9:00|**10:00**|
|TR2  |    |    |    |    |    |9:03|10:02    |
|PT3  |    |    |    |    |    |    |         |

Starting from A at 7:55:

|level| A  | B  | C  | D  | E  | F  |  G      |
|-----|----|----|----|----|----|----|---------|
|TR0  |7:55|    |    |    |    |    |         |
|PT1  |    |8:30|8:33|8:36|9:10|    |         |
|TR1  |    |8:32|8:35|8:38|9:12|    |         |
|PT2  |    |    |    |    |9:00|9:00|**10:20**|
|TR2  |    |    |    |    |9:02|9:03|10:22    |
|PT3  |    |    |    |    |    |    |**10:00**|
|TR3  |    |    |    |    |    |    |10:02    |
|PT4  |    |    |    |    |    |    |         |

> The different PTx on the destination stop points gives the earliest arrival to our destinations for each number of connection. In this case, if we target G, starting at 7:45 from A gives G at 10:00using 0 connection, but starting at 7:55 gives 12:00 using 0 connection and 10:00 using 1 connection.

Reviewer: missing space "at 10:00` `using"

> The different PTx on the destination stop points gives the earliest arrival to our destinations for each number of connection. In this case, if we target G, starting at 7:45 from A gives G at 10:00using 0 connection, but starting at 7:55 gives 12:00 using 0 connection and 10:00 using 1 connection.

Reviewer 2: I don't think I get your point here. Also; that there is no 12:00 stop time in the example.

Reviewer: Description doesn't match the table I think

### Second pass

> As our main objective is earliest arrival time, and then, at equal earliest arrival time, the latest departure time, We need another optimization as RAPTOR doesn't minimize this. Therefore, we do a second pass on the journey: for each candidate arrivals, we rerun RAPTOR in other way around, minimizing departure times.

Reviewer: We kinda maximize departure times.
It could be clearer to say we minimize travel time.
Or maybe just "optimize" ?

> As our main objective is earliest arrival time, and then, at equal earliest arrival time, latest departure time, We need another optimization as RAPTOR doesn't minimize this. For that, we do a second pass on the journey: for each candidate arrivals, we rerun RAPTOR in the other sens, minimizing departure times.

Reviewer 2: ```suggestion
As our main objective is earliest arrival time, and then, at equal earliest arrival time, the latest departure time, We need another optimization as RAPTOR doesn't minimize this. Therefore, we do a second pass on the journey: for each candidate arrivals, we rerun RAPTOR in other way around, minimizing departure times.
```

> The number of second passes can be quite high. To limit this, we compute a bound of the journeys ending to our current second pass initialisation, and check if this bound is dominated by the previously found journeys (found by the previous second passes). Thanks to this optimization, we can avoid a lot of useless second passes. To be safe, we globally bound the number of second passes realized (see the `max_extra_second_pass` parameter of the algorithm).

Reviewer: Maybe here we can say that we only start considering walking at that point.

> The number of second pass can be quite high. For this, we compute a bound of the journeys ending to our current second pass initialisation, and check if this bound is dominated by the previously found journeys (found by the previous second passes). Thanks to this optimization, we can avoid a lot of useless second pass. To be safe, we globally bound the number of second passes realized (see the `max_extra_second_pass` parameter of the algorithm).

Reviewer 2: ```suggestion
The number of second passes can be quite high. To limit this, we compute a bound of the journeys ending to our current second pass initialisation, and check if this bound is dominated by the previously found journeys (found by the previous second passes). Thanks to this optimization, we can avoid a lot of useless second passes. To be safe, we globally bound the number of second passes realized (see the `max_extra_second_pass` parameter of the algorithm).
```

|TR3  |    |    |    |    |    |    |10:02    |
|PT4  |    |    |    |    |    |    |         |

The different PTx on the destination stop points gives the earliest arrival to our destinations for each number of connection. In this case, if we target G, starting at 7:45 from A gives G at 10:00 using 0 connection, but starting at 7:55 gives 10:20 using 0 connection and 10:00 using 1 connection.

> Note that the TR0 can have several entries, and, in practice, there is a few hundred of them (all the accessible stop point by feet).

Reviewer: ```suggestion
Note that the TR0 can have several entries, and, in practice, there is a few hundred of them (all the accessible stop points by feet).
```

## Some ideas for improvement

> If I had to rewrite it, I'd base the algorithm on Connection Scan Algorithm. RAPTOR and CSA are still the simple, flexible and effective algorithms. CSA is simpler, RAPTOR is better when the journey patterns contain a lot of vehicle journeys. Other algorithms try to limit the search space by using the "locality" of the data (in the geographical sense, using contractions or partitionning), but they are much more complicated, much less flexible, and need a complex prepocessing.

Reviewer: Not sure about the first person `I` in doc, but it can be changed later, ideas matter more :wink:

Author: But that's really a personal thought, I can't be impersonal here.

> If I had to rewrite it, I'd base the algorithm on Connection Scan Algorithm. RAPTOR and CSA are still the simple, flexible and effective algorithms. CSA is simpler, RAPTOR is better when the journey patterns contain a lot of vehicle journeys. Other algorithms try to limit the search space by using the "locality" of the data (in the geographical sense, using contractions or partitionning), but they are much more complicated, much less flexible, and need a complex prepocessing.

Reviewer: ```suggestion
If I had to rewrite it, I'd base the algorithm on Connection Scan Algorithm. RAPTOR and CSA are still the simplest, more flexible and more effective algorithms. CSA is simpler, RAPTOR is better when the journey patterns contain a lot of vehicle journeys. Other algorithms try to limit the search space by using the "locality" of the data (in the geographical sense, using contractions or partitionning), but they are much more complicated, much less flexible, and need a complex prepocessing.
```

> As direct objectives, I'd use:

Reviewer: How about price?
And being able to update cost function?

Author: Price is a dedicated area of research. An evolution, not an improvement of the current algorithm.

- minimize arrival time
- minimize number of transfer
- minimize walking duration

> - maximize departure time

Reviewer: This one is "lexicographically" less important than minimizing arrival-time, no?

Author: If you have it as an objective, you have a range query
